db/db_impl_open.cc

```diff
@@ -493,6 +493,9 @@ Status DBImpl::RecoverLogFiles(const std::vector<uint64_t>& log_numbers,
   bool stop_replay_by_wal_filter = false;
   bool stop_replay_for_corruption = false;
   bool flushed = false;
+  // record the maximum sequence number seen among all log files
+  // (excluding skipped content)
+  SequenceNumber max_sequence_all_WALs(0);
```

Comment:
I am pretty sure this will have the same value as `*next_sequence - 1` after the giant loop. A couple side benefits of reusing `next_sequence`: (1) it's initialized to `kMaxSequenceNumber` so you don't need a special case for no WALs anymore; (2) write batches can contain multiple entries, which get different seqnums upon inserting to memtable, so I think it's more correct (sorry, I told you the wrong answer when you asked me earlier whether a WriteBatch can have multiple seqnums).